test/blockchain/blockchain_test.cpp

```diff
@@ -0,0 +1,242 @@
+struct BtcBlockchainInstance {
+  using block_t = BtcBlock;
+  using index_t = typename BlockTree<block_t>::index_t;
+  using height_t = typename BlockTree<block_t>::height_t;
+  using params_t = BtcChainParams;
+
+  std::shared_ptr<StrictMock<BlockRepositoryMock<index_t>>> repo;
+  std::shared_ptr<StrictMock<CursorMock<height_t, index_t>>> cursor;
+  std::shared_ptr<BlockTree<block_t>> blockchain;
+
+  std::shared_ptr<params_t> chainparam;
+  std::shared_ptr<Miner<block_t, params_t>> miner;
+
+  height_t height = 0;
+  ValidationState state;
+
+  BtcBlockchainInstance() {
+    chainparam = std::make_shared<BtcChainParamsRegTest>();
+    miner = std::make_shared<Miner<block_t, params_t>>(chainparam);
+
+    cursor = std::make_shared<StrictMock<CursorMock<height_t, index_t>>>();
+    repo = std::make_shared<StrictMock<BlockRepositoryMock<index_t>>>();
+    blockchain = std::make_shared<BlockTree<block_t>>(repo);
+
+    EXPECT_CALL(*repo, getCursor()).WillRepeatedly(Return(cursor));
+
+    // @given
+    auto genesis = chainparam->getGenesisBlock();
+
+    EXPECT_CALL(*repo, put(Field(&index_t::header, genesis)))
+        .WillRepeatedly(Return(false));
+    EXPECT_CALL(*cursor, seekToFirst()).Times(1);
+    // database is empty, so first isValid will return false
+    EXPECT_CALL(*cursor, isValid()).WillOnce(Return(false));
+
+    // @when
+    auto result = blockchain->bootstrap(height, genesis, state);
+    EXPECT_TRUE(result);
+    EXPECT_TRUE(state.IsValid());
+  };
+};
+
+struct VbkBlockchainInstance {
+  using block_t = VbkBlock;
+  using index_t = typename BlockTree<block_t>::index_t;
+  using height_t = typename BlockTree<block_t>::height_t;
+  using params_t = VbkChainParams;
+
+  std::shared_ptr<StrictMock<BlockRepositoryMock<index_t>>> repo;
+  std::shared_ptr<StrictMock<CursorMock<height_t, index_t>>> cursor;
+  std::shared_ptr<BlockTree<block_t>> blockchain;
+
+  std::shared_ptr<params_t> chainparam;
+  std::shared_ptr<Miner<block_t, params_t>> miner;
+
+  height_t height = 0;
+  ValidationState state;
+
+  VbkBlockchainInstance() {
+    chainparam = std::make_shared<VbkChainParamsRegTest>();
+    miner = std::make_shared<Miner<block_t, params_t>>(chainparam);
+
+    cursor = std::make_shared<StrictMock<CursorMock<height_t, index_t>>>();
+    repo = std::make_shared<StrictMock<BlockRepositoryMock<index_t>>>();
+    blockchain = std::make_shared<BlockTree<block_t>>(repo);
+
+    EXPECT_CALL(*repo, getCursor()).WillRepeatedly(Return(cursor));
+
+    // @given
+    auto genesis = chainparam->getGenesisBlock();
+
+    EXPECT_CALL(*repo, put(Field(&index_t::header, genesis)))
+        .WillRepeatedly(Return(false));
+    EXPECT_CALL(*cursor, seekToFirst()).Times(1);
+    // database is empty, so first isValid will return false
+    EXPECT_CALL(*cursor, isValid()).WillOnce(Return(false));
+
+    // @when
+    auto result = blockchain->bootstrap(height, genesis, state);
+    EXPECT_TRUE(result);
+    EXPECT_TRUE(state.IsValid());
+  };
+};
```

Comment:
these two can be combined into single templated BlockchainInstance, and then parametrized with according types:
```
struct BtcTestCase {
  using params_t = BtcChainParams;
  using hash_t = decltype(BtcBlock::previousBlock);
  using block_t = BtcBlock;
  using height_t = int;
};

struct VbkTestCase {
  using params_t = VbkChainParams;
  using hash_t = decltype(VbkBlock::previousBlock);
  using block_t = VbkBlock;
  using height_t = int;
};
```